[Profiler] Defer recording startup python events

slgong-fb opened the pull request:
Summary:
https://github.com/pytorch/pytorch/issues/81559
- Bypass recording startup calls, instead store their TraceKey and start_time
- Push stored startup calls into 'enters' (set_start_frames)
- Try best-effort inference of system TID for startup calls (populate())

Test Plan:
Lunch a workload that has python modules:
buck build mode/opt kineto/libkineto/fb/integration_tests:pytorch_resnet_integration_test
./buck-out/gen/kineto/libkineto/fb/integration_tests/pytorch_resnet_integration_test.par

Differential Revision: D38029112

slgong-fb commented:
@pytorchbot merge -g

robieta commented on test/profiler/test_profiler.py:
```diff
@@ -1476,6 +1477,85 @@ def test_scalar_ins(self):
+                flat_out_extrafields(node.children, out)
+            return out
+
+        class Counter():
```
Dead code?

slgong-fb replied:
Removed. Thanks!

robieta reviewed (approved):
LGTM! (Modulo tiny nit on unit test.)

slgong-fb committed:
[Profiler] Defer recording startup python events (#82154)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/82154

https://github.com/pytorch/pytorch/issues/81559
- Bypass recording startup calls, instead store their TraceKey and start_time
- Push stored startup calls into 'enters' (set_start_frames)
- Try best-effort inference of system TID for startup calls (populate())

Test Plan:
PROFILER_SOFT_ASSERT_RAISES=1 buck run mode/opt //caffe2/test:profiler -- -r test_startup

buck run mode/opt //caffe2/test:profiler

Reviewed By: robieta

Differential Revision: D38029112

fbshipit-source-id: d0895d545c1a627d9f3ec8c8d82dc314681e713f

dzhulgakov commented:
I think the late assignment of thread ids might be an overkill. CPython seems to have the right ids available in `PyThreadState`: https://github.com/python/cpython/blob/ee821dcd3961efc47262322848267fe398faa4e4/Include/cpython/pystate.h#L159-L165 . If we want an implementation-independent way to get it we could also call out to python and invoke `threading.get_native_id` (https://docs.python.org/3/library/threading.html#threading.get_native_id)
